File: rb/test/currentvector_entity_test.rb
# Currentvector entity test

require "minitest/autorun"
require "json"
require_relative "../YugiLimitRegulation_sdk"
require_relative "runner"

class CurrentvectorEntityTest < Minitest::Test
  def test_create_instance
    testsdk = YugiLimitRegulationSDK.test(nil, nil)
    ent = testsdk.Currentvector(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = currentvector_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "currentvector." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set YUGILIMITREGULATION_TEST_CURRENTVECTOR_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    currentvector_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.currentvector")))
    currentvector_ref01_data = nil
    if currentvector_ref01_data_raw.length > 0
      currentvector_ref01_data = Helpers.to_map(currentvector_ref01_data_raw[0][1])
    end

    # LIST
    currentvector_ref01_ent = client.Currentvector(nil)
    currentvector_ref01_match = {}

    currentvector_ref01_list_result, err = currentvector_ref01_ent.list(currentvector_ref01_match, nil)
    assert_nil err
    assert currentvector_ref01_list_result.is_a?(Array)

  end
end

def currentvector_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "currentvector", "CurrentvectorTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = YugiLimitRegulationSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["currentvector01", "currentvector02", "currentvector03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["YUGILIMITREGULATION_TEST_CURRENTVECTOR_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "YUGILIMITREGULATION_TEST_CURRENTVECTOR_ENTID" => idmap,
    "YUGILIMITREGULATION_TEST_LIVE" => "FALSE",
    "YUGILIMITREGULATION_TEST_EXPLAIN" => "FALSE",
    "YUGILIMITREGULATION_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["YUGILIMITREGULATION_TEST_CURRENTVECTOR_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["YUGILIMITREGULATION_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["YUGILIMITREGULATION_APIKEY"],
      },
      extra || {},
    ])
    client = YugiLimitRegulationSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["YUGILIMITREGULATION_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["YUGILIMITREGULATION_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: lua/test/currentvector_entity_test.lua
-- Currentvector entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("yugi-limit-regulation_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("CurrentvectorEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Currentvector(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["currentvector"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:Currentvector(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:Currentvector(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = currentvector_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "currentvector." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set YUGILIMITREGULATION_TEST_CURRENTVECTOR_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local currentvector_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.currentvector")))
    local currentvector_ref01_data = nil
    if #currentvector_ref01_data_raw > 0 then
      currentvector_ref01_data = helpers.to_map(currentvector_ref01_data_raw[1][2])
    end

    -- LIST
    local currentvector_ref01_ent = client:Currentvector(nil)
    local currentvector_ref01_match = {}

    local currentvector_ref01_list_result, err = currentvector_ref01_ent:list(currentvector_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(currentvector_ref01_list_result)

  end)
end)

function currentvector_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/currentvector/CurrentvectorTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read currentvector test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "currentvector01", "currentvector02", "currentvector03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("YUGILIMITREGULATION_TEST_CURRENTVECTOR_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["YUGILIMITREGULATION_TEST_CURRENTVECTOR_ENTID"] = idmap,
    ["YUGILIMITREGULATION_TEST_LIVE"] = "FALSE",
    ["YUGILIMITREGULATION_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["YUGILIMITREGULATION_TEST_CURRENTVECTOR_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["YUGILIMITREGULATION_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["YUGILIMITREGULATION_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["YUGILIMITREGULATION_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
